fabric-api-lookup-api-v1/src/main/java/net/fabricmc/fabric/impl/lookup/block/BlockApiCacheImpl.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.lookup.block;
+
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.block.BlockState;
+import net.minecraft.block.entity.BlockEntity;
+import net.minecraft.server.world.ServerWorld;
+import net.minecraft.util.math.BlockPos;
+
+import net.fabricmc.fabric.api.event.lifecycle.v1.ServerBlockEntityEvents;
+import net.fabricmc.fabric.api.lookup.v1.block.BlockApiCache;
+import net.fabricmc.fabric.api.lookup.v1.block.BlockApiLookup;
+
+public final class BlockApiCacheImpl<T, C> implements BlockApiCache<T, C> {
+	private final BlockApiLookupImpl<T, C> lookup;
+	private final ServerWorld world;
+	private final BlockPos pos;
+	/**
+	 * We always cache the block entity, even if it's null. We rely on BE load and unload events to invalidate the cache when necessary.
```

Comment:
What happens if you try to cache something that isnt a BE? registred via `registerForBlocks`.

Reply:
If there is no block entity at the target position, the following should happen:
- If the block state is not provided to the cache it will be queried in the world.
- If the provider for the block state is a `WrappedBlockEntityProvider`, which is not supposed to happen, it should probably log an error and not pass a null BE. (I should change that). Otherwise, the provider will receive `(world, pos, state, context)`.
- The fallback provider will receive a `null` block entity instance.
Other than that, the cache should still handle that case correctly.

Not sure if that's the explanation you expected. If needed, I can probably explain better on Discord, just ping me.

Reply:
I really need to boot it up and play with it. Easier to find out that way :)